### ISSUE TITLE
[❔ other question] Optimisation using reference image loaded from PNG.

### ISSUE THREAD

**Issue by JungShinPlusPlus** [mitsuba-renderer/mitsuba2#587]
Hi, I'm experimenting based on cbox example and have a question about the value range (or color space) of rendered image.

In the example the reference is generated by
image_ref = render(scene, spp=spp)
I checked its min, max value and it was "[0, 21.038242]"

after saving the ref image to a PNG, (I'm aware that if save to exr which preserve values without modification but i'm explicitly saving to png atm) I loaded the saved image using,
Bitmap(out_ref_path).convert(Bitmap.PixelFormat.RGB, Struct.Type.Float32, srgb_gamma=False) # tried both srgb_gamma to true and false.
The value range is "[0, 1]", which is expected since it is converted from [0, 255] -> [0, 1.0]

and the optimisation is having hard time because they are in different range. "[0, 21.038242]" vs "[0, 1]"

I'm trying to do with with loading from PNG since the target I'm trying to optimise to cannot be "rendered" for the final application I want to build. (target reference image would some image I get from somewhere.)

So the question is, what is the color space/range of rendered image so I can normalise it to [0, 1.0] to compare with a image that I load from PNG? (converting png image into space of rendered image is ok too. just want them to be in the same range.)

**Comment by Speierers** [comment]
Hi @JungShinPlusPlus,


Currently Mitsuba 2 only provides a high dynamic range film (`hdrfilm`) which will likely produce pixel values greater than `1.0`. You can try to clamp those values to the `[0, 1]` range, knowing that you will "lose" part of the dynamic range here.

Also make sure you convert your loaded PNG images with `srgb_gamma=False` as this will revert the tone mapping baked in those images.

Let me know if this helps!

**Comment by JungShinPlusPlus** [comment]
@Speierers Thank you for the fast feedback!
Yes. clipping the range to [0, 1] and comparing the image resulted correct optimisation (at least got the red wall). 
Thank you.

However, I couldn't find a way to clip the range to [0, 1] in enoki. 
I ended up doing the clipping using pytorch by using render_torch to render and torch.clamp.
But I noticed that if I use pytorch related function it takes much more memory (which is understandable as stated in the document).
Is there a way to do the clip in pure enoki way?

**Comment by Speierers** [comment]
Did you try `ek.clamp(v, min, max)`? or simply using `ek.min()` and `ek.max()`?

**Comment by JungShinPlusPlus** [comment]
Thank you @Speierers. ek.clamp totally worked.

I have a bit of setup issue that I'm using pycharm with setting environment variable to the compiled mistuba.
so the pycharm doesnt does auto completion for mistsuba and ek.
So I tried ek.clip before (just thinking np.clip) and it didnt work so I google a bit for the document but was not successful.
so  I switched to pytorch.

Maybe I'll try to change the setup to make autocompletion work.

Anyway thank you so much for the help. I'll close the ticket :D

**Comment by JakeXin7** [comment]
Hi, I'm facing a similar problem. However, I have set the pixel values all 0, and srgb_gamma=False. Why did the warning occur?

**Comment by JakeXin7** [comment]
I think I get what's wrong with it!
I got the warning :

WARN  main  [BitmapTextureImpl_3_0] BitmapTexture: texture named "black.png" contains pixels that exceed the [0, 1] range!

while optimizing the texture "black.png".
I think this warning occurred because the pixel values exceeded the [0, 1] range during the optimization.
So how can I ignore the warning? This is a common situation while optimizing the texture, hence this waring is not necessary but a waste of time in this case.

**Comment by Speierers** [comment]
For this you can you the 'ek::clamp' as mentioned above

**Comment by JakeXin7** [comment]
OK, I'll try this method. Thanks for reply.
